### PR TITLE
[WIP] ENH Adds tree reduction to bincount

### LINKED ISSUE #4852: Use tree reduction in `da.bincount`
Currently `da.bincount` does a [reduction over `np.bincount`s from each chunk]( https://github.com/dask/dask/blob/2f308e72a46de818055819108429e8f7ce7ba03b/dask/array/routines.py#L549 ). It would be nice to change this to use [a tree reduction]( https://github.com/dask/dask/blob/master/dask/array/reductions.py ) instead.

### PR DESCRIPTION
**Pull request by thomasjpfan**
- [x] Closes #4852
- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

This implementation is not faster with this benchmark:

```py
import dask.array as da
import numpy as np

rng = np.random.RandomState(42)
x = rng.randint(5000, size=10_000_000)
x_da = da.from_array(x, chunks=300_000)
out = da.bincount(x_da)
```

```py
%%timeit
_ = out.compute()
```

### This PR

```
26.1 ms ± 2.06 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

### Master

```
21.2 ms ± 272 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Maybe there is not a benefit to tree reduction here?

### CODE REVIEW

**Comment by jsignell** [comment]
Huh. If there is no improvement do you think it's worth including this? Or do you think there'd be an improvement with a different benchmark?

**Comment by thomasjpfan** [comment]
I think the benefit of this PR is memory usage for larger arrays:

```py
import dask.array as da
import dask
from dask.diagnostics import Profiler, ResourceProfiler, CacheProfiler
import numpy as np

rng = np.random.RandomState(42)
x = rng.randint(50000, size=500_000_000)
x_da = da.from_array(x, chunks=100_000)
out = da.bincount(x_da)

with ResourceProfiler(dt=0.01) as rprof:
    out.compute()
    
rprof.visualize()
```

On `master`: Goes to ~ 7.4 GB

bokeh_plot-2

This PR: Stays at ~ 4.7GB

bokeh_plot

`master` is always a little faster but it uses more memory.

**Comment by jsignell** [comment]
Oh that is a huge memory improvement! Thanks for taking the time to write that up :)

File: dask/array/routines.py [hunk 1]
```diff
@@ -629,43 +632,38 @@ def _bincount_sum(bincounts, dtype=int):
...
-    final_name = "bincount-sum" + token
-    # Call np.bincount on each block, possibly with weights
+    args = [x, "i"]
     if weights is not None:
-        dsk = {
-            (name, i): (np.bincount, (x.name, i), (weights.name, i), minlength)
-            for i, _ in enumerate(x.__dask_keys__())
-        }
-        dtype = np.bincount([1], weights=[1]).dtype
+        meta = meta_from_array(weights)
```
Comment (jsignell) [inline dask/array/routines.py:645]: I think the meta should be taken from some small version of the `bincount` like how dtype was before:

```suggestion
        meta = np.bincount([1], weights=[1])
```

Here's a test for this case:

```python
def test_bincount_with_int_weights():
    x = np.array([2, 1, 5, 2, 1])
    d = da.from_array(x, chunks=2)
    weights = np.array([1, 2, 1, 0, 1])

    dweights = da.from_array(weights, chunks=2)
    e = da.bincount(d, weights=dweights, minlength=6)
    assert_eq(e, np.bincount(x, weights=dweights.compute(), minlength=6))
    assert same_keys(da.bincount(d, weights=dweights, minlength=6), e)
```

File: dask/array/tests/test_routines.py [hunk 2]
```diff
@@ -535,10 +535,16 @@ def test_bincount():
...
 
 
-def test_bincount_with_weights():
+@pytest.mark.parametrize(
+    "weights",
+    [
+        np.array([1, 2, 1, 0.5, 1], dtype=np.float32),
+        np.array([1, 2, 1, 0, 1], dtype=np.int32),
+    ],
+)
```
Comment (jsignell) [inline dask/array/tests/test_routines.py:544]: :clap:

**Comment by thomasjpfan** [comment]
Same benchmark with `Profiler` included:

## On `master`

Screen Shot 2021-02-23 at 3 56 11 PM

## This PR

Screen Shot 2021-02-23 at 3 55 41 PM

My guess is that there is more gil contention when running `_bincount_agg` and `np.bincount` together.

**Comment by jsignell** [comment]
This seems reasonable to me. Thanks for doing this work!

**Comment by jakirkham** [comment]
Thanks Thomas! 😀

**Pull request merged**